mesonbuild/utils/core.py

```diff
@@ -68,6 +68,9 @@ class HoldableObject(metaclass=abc.ABCMeta):
     ''' Dummy base class for all objects that can be
         held by an interpreter.baseobjects.ObjectHolder '''
 
+    def display_repr(self) -> str:
+        return '--internal--'
```

Comment:
This doesn't seem like a very useful value to display even for debugging purposes.

Reply:
I was following [xclaesse's advice](https://github.com/mesonbuild/meson/issues/8207#issuecomment-761023733). How about returning the typename here?

Reply:
Maybe base class should raise `NotImplemented`

Reply:
> Maybe base class should raise `NotImplemented`

That's a bad idea. `message([1,2,3, m_dep])` should print `[1, 2, 3, something]`.

Reply:
What are the subclasses that don't implement it?

Reply:
Currently, only  `file`, `compiler`, `module` and `disabler` objects implement it.

Reply:
We could also do this:
```
def display_repr(self) -> str:
        return type(self).__name__ + repr(self.__dict__)
```
Pros:
    - Simple and robust to write. 
    - User gets all the info he needs
Cons:
    - Will not be stable
    - User sees a lot of (potentially uninteresting) internal stuff

Reply:
Stability isn't so worrying, since `message()` cannot affect the build, only print status reports about it to the console. If we change the representation value of complex types, nothing will break.

We could always document that complex types will produce a human readable message that isn't guaranteed to be machine-readable.

Reply:
At least the type `__name__` seems practical to print, unsure about the `__dict__`...